velox/common/caching/StringIdMap.cpp

```diff
@@ -43,13 +43,126 @@ void StringIdMap::release(uint64_t id) {
   }
 }
 
+#define VELOX_STRINGIDMAP_PREFIX "[StringIdMap] "
+#define VELOX_STRINGIDMAP_ERROR_LOG LOG(ERROR) << VELOX_STRINGIDMAP_PREFIX
+
+void StringIdMap::checkAndReportIntegrity(uint64_t id) const {
+  if (idToString_.size() != stringToId_.size()) {
+    VELOX_STRINGIDMAP_ERROR_LOG << fmt::format(
+        "idToString size {}. stringToId size {}. ",
+        idToString_.size(),
+        stringToId_.size());
+  }
+
+  if (idToString_.size() != lastId_) {
+    VELOX_STRINGIDMAP_ERROR_LOG << fmt::format(
+        "idToString {} element(s) were erased. ", lastId_ - idToString_.size());
+  }
+
+  if (stringToId_.size() != lastId_) {
+    VELOX_STRINGIDMAP_ERROR_LOG << fmt::format(
+        "stringToId {} elements were erased. ", lastId_ - stringToId_.size());
+  }
+
+  std::vector<std::pair<uint64_t, std::string>> missingFiles;
+  uint64_t pinnedSize{0};
+  for (auto it = stringToId_.begin(); it != stringToId_.end(); ++it) {
+    pinnedSize += it->first.length();
+    if (it->second == id) {
+      VELOX_STRINGIDMAP_ERROR_LOG << fmt::format(
+          "Found missing id {} in stringToId with file \"{}\".", id, it->first);
+    }
+    auto idEntry = idToString_.find(it->second);
+    if (idEntry == idToString_.end()) {
+      missingFiles.push_back(
+          std::pair<uint64_t, std::string>(it->second, it->first));
+    }
+  }
+
+  if (pinnedSize != pinnedSize_) {
+    VELOX_STRINGIDMAP_ERROR_LOG << fmt::format(
+        "stringToId pinnedSize {} does not match actual pinnedSize {}.",
+        pinnedSize_,
+        pinnedSize);
+  }
+
+  for (auto fileIt = missingFiles.begin(); fileIt != missingFiles.end();
+       ++fileIt) {
+    if (fileIt->first != id) {
```

Comment:
I didn't follow this check either. Under what conditions could we hit this ?

Reply:
The missingFiles list indicates which files are in `stringToId_` but have no corresponding id entry in `idToString_`. We already know that `id` wasn't in `idToString_` hence why we come here. So we only want to print ids that are additionally not found in case they exist.
Basically, the check is here skip printing that `id` was not found. Of course, this is assuming that `id` did have an entry in `stringToId_` still - which it may or may not have.